Add Unit Tests for mtbl v1.7.0 Features

kadealicious opened the pull request:
- Add a unit test for `-t` multithreading CLI option.
- Add a unit test for DNS and DNSSEC `-b` block size CLI option.
- Add a common long-parsing function to ensure consistency between CLI options.

kadealicious committed:
Add unit tests for mtbl 1.7.0 threading and blocksize options

kadealicious committed:
Improve command-line option parsing with parse_long()

shw700 commented on dnstable_convert.c:
```diff
@@ -1017,6 +1017,26 @@ init_mtbl(mtbl_compression_type compression, int level, size_t dns_block_size,
+		return false;
+	}
+
+	if (endptr == str || *endptr != '\0')
```
I think the latter half of the comparison suffices... `endptr == str` is really just a sub-case of `*endptr != '\0'`

Could even become `return (*endptr != '\0');`

kadealicious replied:
Good point; fixed!

shw700 commented on dnstable_convert.c:
```diff
@@ -1101,24 +1121,21 @@ main(int argc, char **argv)
 		case 'l':
-			compression_level = strtol(optarg, &end, 10);
-			if (*end != '\0') {
+			if (!parse_long(optarg, &compression_level) || compression_level < 0) {
```
See `mtbl` PR #89 for comments on this `compression_level` check and `thread_count` below...

kadealicious replied:
See response here: https://github.com/farsightsec/mtbl/pull/89#discussion_r1889077802

kadealicious replied:
Fixed

shw700 commented:
Will still need to dispose of tmp files after unit test completion.

reedjc commented:
I committed two small changes.

Some more comments:

1) Maybe check the output of the dnstable_convert for the new tests too. The output is collected in your tests but not used.
2) Maybe use mtbl_info (like in the mtbl tests) to look for the block sizes.
3) Note that -b 256,256 is below the minimums MIN_BLOCK_SIZE allowed so they are silently increased. We could check for that "data block size" specifically.
4) Currently the Makefile.am removes all the tests outputs which is redundant to the new "rm" at the end of the test script. I didn't remove the test outputs here by default before so i could troubleshoot failures. So now mix of ways to remove files is inconsistent.
5) And if using data to troubleshoot, don't remove/reuse the $new_dns_blocksize / $dns_blocksize files but use new filenames for different tests?
6) The specific rm for one file $unordered_rdata_dns is redundant to previous rm of the wildcard.
7) instead of using "new_" as filename prefix just exclude it and then the other variables prefix to "default_" ? (As "new" won't mean "new" soon :)

Thank your for all your tests. Even without considering any of the above thoughts, I am fine with these changes as-is and you may merge.

shw700 commented on tests/tests.sh.in:
```diff
@@ -245,5 +251,77 @@ else
+# work as intended.
+
+# Verify that dnstable_convert output does not change when the -t option is specified.
+echo -n "verify dnstable_convert -t option does not change output: "
```
The input to this consists of only two nmsg payloads, no? This doesn't seem to be a sufficient candidate for multi-threaded testing... and 10 threads to process two total entries seems like extreme overkill.

kadealicious replied:
Added a much larger file

reedjc commented:
"As of writing this, mtbl_info does not output DNSSEC block size". There is no difference between "dns" vs dnssec for the block sizes. Just use the same mtbl info which has nothing to do with dns/dnssec:

for example:

$ ~/opt/mtbl/bin/mtbl_info ./tests/blocksize1.dnssec.mtbl.tmp ./tests/default_blocksize1.dnssec.mtbl.tmp | egrep 'file | block size'
file name:             ./tests/blocksize1.dnssec.mtbl.tmp
file size:             744
data block size:       12345
file name:             ./tests/default_blocksize1.dnssec.mtbl.tmp
file size:             744
data block size:       65536

reedjc commented:
I committed some changes to the branch.
Also see my earlier comment about testing the blocksize (I can handle the code changes if you want).

Lastly see the DISTCLEANFILES. We don't have to do it that way, but currently around 16 files are left around after a make distclean.   We can do this simpler by using globbing wildcards for the filenames to remove in the DISTCLEANFILES and changing a few of the generated filenames to have consistent output naming so we can remove them using wildcards (without removing source files). (I can handle the code changes if you want in this same branch.)

kadealicious commented:
> I committed some changes to the branch. Also see my earlier comment about testing the blocksize (I can handle the code changes if you want).
> 
> Lastly see the DISTCLEANFILES. We don't have to do it that way, but currently around 16 files are left around after a make distclean. We can do this simpler by using globbing wildcards for the filenames to remove in the DISTCLEANFILES and changing a few of the generated filenames to have consistent output naming so we can remove them using wildcards (without removing source files). (I can handle the code changes if you want in this same branch.)

Fixed block size test to use `mtbl_info` (apologies for the brain fart), renamed temporary test files to *.tmp, and added `tests/*.tmp` to `DISTCLEANFILES` (also addressing [this](https://github.com/farsightsec/dnstable-convert/pull/23#issuecomment-2552315570) comment).

reedjc commented:
See the additional DISTCLEANFILES in the Makefile.am. Maybe entirely remove it if became redundant or wrong.

shw700 commented on tests/tests.sh.in:
```diff
@@ -245,5 +252,79 @@ else
+$DNSTABLE_CONVERT @abs_srcdir@/dnssec.nmsg $default_dns_blocksize1 $default_dnssec_blocksize1 2>&1 | strip_env_info | sort -f > @abs_builddir@/dnstable_convert.output9.tmp
+default_blocksize1=$($MTBL_INFO $default_dnssec_blocksize1 | awk '/^data block size:/{print $4}')
+blocksize1=$($MTBL_INFO $set_dnssec_blocksize1 | awk '/^data block size:/{print $4}')
+cmp -s $set_dnssec_blocksize1 $default_dnssec_blocksize1
```
I think the most robust test here would be to compare the awk-extracted block sizes (as is currently being done), and then use `mtbl_dump` to verify that the `mtbl` key/value pairs ARE identical.

Of course the raw files will differ; otherwise they would not be able to advertise different data block sizes in their headers.

kadealicious replied:
Good idea; fixed for both DNS and DNSSEC block size tests.
